File phone restore implementation

benrudolph opened the pull request:
@snopoke @czue This is the initial implementation of file phone restore. I've tested it by modifying the existing tests that test `get_payload` and setting the proper toggle. I didn't check the tests in yet. I wanted to get feedback on it thus far before diving in further.
- [x] Implement phone restore using files
- [x] Ensure all places that call `get_payload` expect a string response or filename (https://github.com/benrudolph/commcare-hq/commit/b7b08e73acd4bb7ddb55069994f66f9ca777b3ba)
- [x] Cron job for clearing `restore_payloads` directory (https://github.com/dimagi/commcarehq-ansible/pull/58)
- [x] Test coverage (https://github.com/benrudolph/commcare-hq/commit/aa7f2184ed3e7a4f73a4efbbd0296edf7e82f358 & https://github.com/benrudolph/commcare-hq/commit/4b390a02186589b5f2b4dae19e221bddd3648fbe )

It's pretty clunky trying to wrap the response body in an opening and closing tag which is why I have to create two files and do a copy. I've been using `shutil.copyfileobj` which seems like the best option to do concatenation after googling around a bit, but if you know a better way I'd be all ears.

:leopard:

benrudolph committed:
File Phone Restore implementation

snopoke commented:
Regarding cleaning out old files you don't need a management command. You can do it in a one line cron job e.g. https://github.com/dimagi/commcarehq-ansible/blob/master/ansible/roles/touchforms/tasks/main.yml#L32-32

snopoke commented on corehq/ex-submodules/casexml/apps/phone/restore.py:
```diff
@@ -93,17 +97,72 @@ def extend(self, iterable):
+        other.response_body.seek(0)
+
+        shutil.copyfileobj(self.response_body, response.response_body)
+        shutil.copyfileobj(other.response_body, response.response_body)
```
This is the best way to do this but it adds a lot of IO to the process. I was thinking we could get around it but I actually think we're going to need it later when we start doing async restores.

benrudolph replied:
I was wondering about that as well. I tried to touch the algorithm as little as possible, but it seems like we could add a parameter to the `payload_fn` that optionally takes in an existing `RestoreResponse`. That way we don't have to add the two together here: https://github.com/benrudolph/commcare-hq/blob/ota-file-161809/corehq/ex-submodules/casexml/apps/phone/restore.py#L451

snopoke commented:
Quick review looks good. Will come back to this later for a more detailed look.

benrudolph commented:
that's a nice little file for setting up cronjobs, thanks. i just have to make sure that i have access to the `RESTORE_PAYLOAD_DIR` in that file. i was also thinking so i don't accidentally blow away a file while it's being written to, i'd only delete those that were like a few hours old or something

czue commented:
that command can be used to arbitrarily specify how old a file should be before being deleted

benrudolph commented:
ah got it thanks. didn't catch that

benrudolph commented:
just braindumping. if the file gets removed from the `restore_payloads` folder and the filename is still valid in the cache, then we'll definitely hit a 500 somewhere. need to someone make sure the file exists when checking the cache

benrudolph committed:
Replace relevant calls to get_payload

benrudolph committed:
Use StreamingHttpResponse to return file

benrudolph committed:
Update test to use FileResponse as well

dimagimon commented:
There are 7 errors:

* corehq/ex-submodules/casexml/apps/phone/restore.py, line 30 - F401 'get_payload_content' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/test_state_hash.py, line 85 - F821 undefined name 'StateFileHashTest'
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 5 - F401 'clear_toggle_cache' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 5 - F401 'update_toggle_cache' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 9 - F401 'FILE_RESTORE' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 24 - F811 redefinition of unused 'get_payload_content' from line 18
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 935 - W291 trailing whitespace

benrudolph committed:
fix dimagimon errors

benrudolph committed:
Add FileRestoreSyncTokenUpdateTest

benrudolph committed:
Resolve conflicts

benrudolph commented:
(test failure certainly seems real, going to look into it)

benrudolph commented on corehq/ex-submodules/casexml/apps/phone/restore.py:
```diff
@@ -43,6 +48,9 @@ def add_custom_parameter(key, value):
 INITIAL_SYNC_CACHE_THRESHOLD = 60  # 1 minute
 
+# Max amount of bytes to have in memory when streaming a file
+MAX_BYTES = 1000000  # 1MB
```
Don't really have a good idea for how much of the file we should read into memory. Any insight would be helpful

snopoke replied:
Seems reasonable though I don't think we'd have any issue if we increased it to 10Mb which would mean that most responses would be read in a single read.

snopoke commented on corehq/ex-submodules/casexml/apps/phone/restore.py:
```diff
@@ -218,7 +303,20 @@ def get_case_payload(domain, stock_settings, version, user, last_sync, synclog):
+            xml.get_case_element(op.case, op.required_updates, version)
+            for op in batch.case_updates_to_sync()
+        )
+        response.extend(case_xml_elements)
```
Seems like this section has been duplicated.

benrudolph replied:
that would explain the test error. good catch, thanks. i think this happened when i was merging conflicts

snopoke commented on corehq/ex-submodules/casexml/apps/phone/restore.py:
```diff
@@ -381,7 +478,13 @@ def get_payload(self):
+                f = open(payload, 'r')
+                # Since payload file is all one line, need to readline based on bytes
+                return StreamingHttpResponse(iter(lambda: f.readline(MAX_BYTES), ''), mimetype="text/xml")
+            except IOError:
```
what if a real IOError occurs here?

benrudolph replied:
Unclear. this is actually a problem whereever I use that idiom. another solution could be to check if the payload has a `.xml` file extension since all the filenames should have that and none of the XML payload files should. That way we can still guard against an actual `IOError` in which case we might send back a 500 error with an appropriate error message

snopoke replied:
I think this is the only place where this idiom is used outside of test
cases. One option would be to do something like this:
```
try:
    with open(payload, 'r') as f:
        try:
            ....
        exception IOError:
            # do something
except IOError:
    # no such file or directory
```

Another option is to check the `errno` attrib of the exception which should
be [ENOENT](https://docs.python.org/2/library/errno.html#errno.ENOENT) if
the file does not exist. I'm not sure if that's a good practice to do.

benrudolph replied:
ill check on the `ENOENT` seems promising. I'm not sure how that nested try catch would work. How does that differentiate between the two types of IO errors?

snopoke replied:
One is inside the context manager, the other is outside.

benrudolph committed:
Minor improvements

- Use predictably random toggle
- check if payload endswith .xml
- increase max bytes
- a few better comments

benrudolph committed:
Better error handling

benrudolph commented:
Ok I think this is good to go. Just so you know I changed the toggle a few commits back to make it a predictably random one. 5% of users will be using this feature

snopoke commented:
Does this cache the filename or the file contents? https://github.com/dimagi/commcare-hq/pull/5956/files#diff-668e726338b412d3a9f2631438979a57R463

If its just the filename then we should check that the file still exists before returning it.

benrudolph commented:
It's just the filename. Yea good call, I was just hoping that the cache would never be set longer than the removal of the file, but that seems like a bad idea

benrudolph committed:
Failing test to invalidate cache after file has been removed

benrudolph committed:
Fix for invalidating cache when file is deleted

benrudolph commented:
Ok test and fix committed. I couldn't think of a better way (barring trying to open the file) to tell if the payload was a file or string response beyond checking the file extension

dannyroberts commented:
(haven't review the PR, but from the perspective of coordinating with the ansible change, this can be merged whenever y'all see fit)

benrudolph committed:
Resolve toggle conflict

snopoke commented:
Kicked build

czue commented on settings.py:
```diff
@@ -496,6 +496,9 @@
 }
 
+# Note: if you change this, make sure to update the cronjob in commcarehq-ansible
+RESTORE_PAYLOAD_DIR = os.path.join(FILEPATH, 'restore_payloads')
```
how do we feel about putting these in the code root directory? it somehow feels wrong to me, though i don't think i can completely articulate why other than from some abstract "separation of concerns" perspective. @snopoke @benrudolph @dannyroberts

dannyroberts replied:
Agree we generally treat code_root as something we can blow away and replace with a fresh clone. The fact that it's .gitignore'd means that it'll survive the `git clean -ffd` we run on deploy, but still I agree it'd be nice to save somewhere else. We could go off whatever convention we set with touchforms sessions (though I guess those are being moved to a DB). Looking at `touchforms` role in ansible it's: `{{ www_home }}/data/touchforms`.

To do that you'd have to add something analogous in ansible for creating the `{{ www_home }}/data/restore_payloads` directory (like https://github.com/dimagi/commcarehq-ansible/blob/3bf708a6be2830f67abca9d2fbab549c00abef54/ansible/roles/touchforms/tasks/main.yml#L30 for touchforms)

snopoke replied:
+1 on following the touchforms convention

benrudolph replied:
This makes sense to me. `restore_payloads` in the root dir felt wrong to me as well, but i didn't know where else to put it. Looking into making this change

benrudolph replied:
Ok, took this https://github.com/dimagi/commcare-hq/commit/96d8380783b66b0b2af500347fa4b1e0d0dd1dc4 from the touchforms example. This should be all the change that's needed for this PR. I have to modify the ansible script to make the appropriate directory and set the appropriate variable

czue commented:
this looks great to me and i love the tests. i had one pretty minor comment but am happy to merge if no one else thinks it matters

benrudolph committed:
Use tempfile dir

benrudolph committed:
Remove restore_payloads dir

dannyroberts commented:
I think basically as soon as you deploy https://github.com/dimagi/commcarehq-ansible/pull/59, this is good to merge. Feel free to self merge or else ping me and I can merge!

benrudolph commented:
This isn't actually done, in conversation with @czue during the ansible deploy, he realized that since there are 3 django machines, that by caching only the filename we'll be missing caches on 2/3 of the requests. Going to add another commit to stream the file to the cache (may or may not be a significant change)

benrudolph commented:
Alright on further analysis this will be a pretty big change (i know i said the opposite 5 mins ago to you cory). It's simple enough to write the response to the cache the same way i stream it. The issue is, when I get the response back from the cache i'll have to load the entire response into memory, defeating the purpose of the PR?  Not sure how to get around this since redis doesn't really operate as a python IO object and after looking through the docs i don't think it's possible to incrementally read a value from the redis cache. some outlandish ideas: shared filesystem for a certain directory (seems kind of ridiculous for this feature), use postgres to serve as cache (seems like that could go wrong in lots of ways)

benrudolph commented:
One solution might be to create multiple cache keys to store the response in separate pieces, but this also seems like it'll be fragile

snopoke commented:
I think most of them memory issues have been happening on the celery machine when we're trying to generate the response asynchronously using the 'prime restore' feature. In future we'll be doing much more async restores so hence why this is important. I think it's ok to just read it into memory.

In future if we want to cache it as separate chunks I think we've got all the bits necessary, we just need to tie it all together.

benrudolph committed:
Write file contents to cache

benrudolph committed:
Add documentation for get_payload

benrudolph commented:
:white_check_mark:

benrudolph committed:
Fix cache setting

benrudolph committed:
Revert "Fix cache setting"

This reverts commit 4c3c0545be72d65a804b045e17f0d0fe9b77b459.

benrudolph committed:
Revert "Write file contents to cache"

This reverts commit c2c6c20ae9cd4ed8f93c67d9f3de01800ae73a80.

benrudolph committed:
Update documentation

benrudolph committed:
Do not use predictably random toggle

benrudolph committed:
Resolve conflicts

benrudolph committed:
Fix parameter from bad merge

benrudolph commented:
:white_check_mark:

benrudolph commented:
(just to note, i took out the PredictablyRandomToggle and changed back to a regular toggle, so it's a bit safer. I also reverted all the commits I made to have the cache return the file contents rather than the filename)

dannyroberts commented:
@benrudolph is this all set to merge? Have you figured out how you're going to coordinate the deploy with actually setting up the shared filesystem on prod?

benrudolph commented:
I think this is ready to go. Nothing will happen unless someone turns on the toggle, so we can take our time with setting it up on prod. Basically i just need to run the shared_dir role on prod. Looked at the india inventory for a moment today, but didn't make any changes yet

dannyroberts commented:
:+1:
